File: viewer/TweetViewer.Tests/AscendingOrderTests.cs
using System.IO;
using System.Text;
using TweetViewer.Data;
using TweetViewer.Models;

namespace TweetViewer.Tests;

/// <summary>
/// 「古い順」(ascending) 表示のリポジトリ層テスト。
/// 日付は ISO 形式 (MergedTimelineTests の "Tue Jul..." 形式は曜日不一致でパースされず
/// sort_key = 0 に落ちるため、並び順のテストには使えない)。
/// </summary>
public sealed class AscendingOrderTests : IDisposable
{
    private readonly string _dataDir;
    private readonly ViewerDatabase _db;

    public AscendingOrderTests()
    {
        _dataDir = Path.Combine(Path.GetTempPath(), "SubakoTests", Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(_dataDir);
        _db = new ViewerDatabase(_dataDir);
        _db.EnsureCreated();
    }

    public void Dispose()
    {
        Microsoft.Data.Sqlite.SqliteConnection.ClearAllPools();
        try
        {
            Directory.Delete(_dataDir, recursive: true);
        }
        catch (IOException)
        {
        }
    }

    private static string TweetLine(long id, string date) =>
        $$"""{"id":"{{id}}","created_at":"{{date}}","full_text":"tweet {{id}}","user":{"username":"author{{id}}","display_name":"Author {{id}}"},"entities":[]}""";

    private async Task ImportAsync(string username, params string[] lines)
    {
        await new UserRepository(_db).AddAsync(username);
        Directory.CreateDirectory(_db.UserDir(username));
        File.AppendAllText(_db.JsonlPath(username),
            string.Join("", lines.Select(l => l + "\n")), new UTF8Encoding(false));
        await new JsonlImporter(_db).ImportUserAsync(username);
    }

    private static string Date(int day) => $"2026-07-{day:00}T20:00:00+00:00";

    private static long Epoch(int day) =>
        new DateTimeOffset(2026, 7, day, 20, 0, 0, TimeSpan.Zero).ToUnixTimeSeconds();

    [Fact]
    public async Task AscendingReturnsOldestFirst()
    {
        await ImportAsync("alice",
            TweetLine(2, Date(2)), TweetLine(1, Date(1)), TweetLine(3, Date(3)));

        var page = await new TweetRepository(_db).GetPageAsync(
            ["alice"], false, null, 10, ascending: true);

        Assert.Equal(new[] { "1", "2", "3" }, page.Rows.Select(r => r.TweetId));
    }

    [Fact]
    public async Task AscendingKeysetPagesConcatenateWithoutGapsOrDuplicates()
    {
        const string bucket = "searches/kw-x";
        // 重複ツイートがページ境界を跨ぐ構成 (MergedTimelineTests の ASC 版)
        await ImportAsync("alice",
            TweetLine(1, Date(1)), TweetLine(3, Date(3)), TweetLine(5, Date(5)));
        await ImportAsync(bucket,
            TweetLine(2, Date(2)), TweetLine(3, Date(3)), TweetLine(4, Date(4)), TweetLine(5, Date(5)));

        var repo = new TweetRepository(_db);
        var collected = new List<string>();
        (long, long)? cursor = null;
        for (var i = 0; i < 10; i++)
        {
            var page = await repo.GetPageAsync(["alice", bucket], false, cursor, 2, ascending: true);
            collected.AddRange(page.Rows.Select(r => r.TweetId));
            if (page.Rows.Count < 2)
                break;
            var last = page.Rows[^1];
            cursor = (last.SortKey, last.IdInt);
        }

        Assert.Equal(new[] { "1", "2", "3", "4", "5" }, collected);
    }

    [Fact]
    public async Task AscendingKeepsDedupRepresentative()
    {
        const string bucket = "searches/kw-x";
        var shared = TweetLine(2, Date(2));
        await ImportAsync("alice", shared);
        await ImportAsync(bucket, shared, TweetLine(1, Date(1)));

        var page = await new TweetRepository(_db).GetPageAsync(
            ["alice", bucket], false, null, 10, ascending: true);

        Assert.Equal(new[] { "1", "2" }, page.Rows.Select(r => r.TweetId));
        Assert.Equal("alice", page.Rows.Single(r => r.TweetId == "2").Username);   // 実アーカイブが代表
    }

    [Fact]
    public async Task AscendingMediaPageKeepsIdxOrderAndCursorContinues()
    {
        var two = $$"""{"id":"2","created_at":"{{Date(2)}}","full_text":"two","user":{"username":"alice"},"entities":[{"type":"photo","link":"https://pbs.twimg.com/media/A1.jpg"},{"type":"photo","link":"https://pbs.twimg.com/media/A2.jpg"}]}""";
        var one = $$"""{"id":"1","created_at":"{{Date(1)}}","full_text":"one","user":{"username":"alice"},"entities":[{"type":"photo","link":"https://pbs.twimg.com/media/B1.jpg"}]}""";
        await ImportAsync("alice", two, one);
        var repo = new TweetRepository(_db);

        var page = await repo.GetMediaPageAsync(["alice"], null, 10, ascending: true);
        // ツイートは古い順、同一ツイート内の idx は昇順のまま
        Assert.Equal(new[] { ("1", 1), ("2", 1), ("2", 2) },
            page.Select(m => (m.TweetId, m.Idx)));

        // 3 要素カーソルの継続
        var p1 = await repo.GetMediaPageAsync(["alice"], null, 2, ascending: true);
        var last = p1[^1];
        var p2 = await repo.GetMediaPageAsync(
            ["alice"], (last.SortKey, last.IdInt, last.Idx), 2, ascending: true);
        Assert.Equal(new[] { ("1", 1), ("2", 1) }, p1.Select(m => (m.TweetId, m.Idx)));
        Assert.Equal(new[] { ("2", 2) }, p2.Select(m => (m.TweetId, m.Idx)));
    }

    [Fact]
    public async Task AscendingCombinesWithRangeAndUnreadOnly()
    {
        await ImportAsync("alice",
            TweetLine(1, Date(1)), TweetLine(2, Date(2)),
            TweetLine(3, Date(3)), TweetLine(4, Date(4)));
        var repo = new TweetRepository(_db);
        await repo.SetReadAsync("3", "alice", read: true);

        // 期間 [Day2, Day5) かつ未読のみ → 2, 4 が古い順
        var page = await repo.GetPageAsync(
            ["alice"], unreadOnly: true, null, 10,
            new DateRangeFilter(Epoch(2), Epoch(5)), ascending: true);

        Assert.Equal(new[] { "2", "4" }, page.Rows.Select(r => r.TweetId));
    }
}

File: mac/Tests/SubakoCoreTests/AscendingOrderTests.swift
import Foundation
import Testing
@testable import SubakoCore

/// 「古い順」(ascending) 表示のリポジトリ層テスト (C# AscendingOrderTests.cs の移植)。
@Suite struct AscendingOrderTests {
    private let bucket = "searches/kw-x"

    @Test func ascendingReturnsOldestFirst() async throws {
        let t = try TestDataDir()
        defer { t.cleanup() }
        try await t.importUser("alice", [
            tweetLine(2, date: isoDate(2)), tweetLine(1, date: isoDate(1)), tweetLine(3, date: isoDate(3)),
        ])

        let page = try await TweetRepository(t.db).getPage(
            usernames: ["alice"], unreadOnly: false, after: nil, limit: 10, ascending: true)

        #expect(page.rows.map(\.tweetId) == ["1", "2", "3"])
    }

    @Test func ascendingKeysetPagesConcatenateWithoutGapsOrDuplicates() async throws {
        let t = try TestDataDir()
        defer { t.cleanup() }
        // 重複ツイートがページ境界を跨ぐ構成 (MergedTimelineTests の ASC 版)
        try await t.importUser("alice", [
            tweetLine(1, date: isoDate(1)), tweetLine(3, date: isoDate(3)), tweetLine(5, date: isoDate(5)),
        ])
        try await t.importUser(bucket, [
            tweetLine(2, date: isoDate(2)), tweetLine(3, date: isoDate(3)),
            tweetLine(4, date: isoDate(4)), tweetLine(5, date: isoDate(5)),
        ])

        let repo = TweetRepository(t.db)
        var collected: [String] = []
        var cursor: TweetCursor?
        for _ in 0..<10 {
            let page = try await repo.getPage(
                usernames: ["alice", bucket], unreadOnly: false, after: cursor, limit: 2,
                ascending: true)
            collected.append(contentsOf: page.rows.map(\.tweetId))
            if page.rows.count < 2 { break }
            let last = page.rows.last!
            cursor = TweetCursor(sortKey: last.sortKey, idInt: last.idInt)
        }

        #expect(collected == ["1", "2", "3", "4", "5"])
    }

    @Test func ascendingKeepsDedupRepresentative() async throws {
        let t = try TestDataDir()
        defer { t.cleanup() }
        let shared = tweetLine(2, date: isoDate(2))
        try await t.importUser("alice", [shared])
        try await t.importUser(bucket, [shared, tweetLine(1, date: isoDate(1))])

        let page = try await TweetRepository(t.db).getPage(
            usernames: ["alice", bucket], unreadOnly: false, after: nil, limit: 10,
            ascending: true)

        #expect(page.rows.map(\.tweetId) == ["1", "2"])
        #expect(page.rows.first { $0.tweetId == "2" }?.username == "alice")   // 実アーカイブが代表
    }

    @Test func ascendingMediaPageKeepsIdxOrderAndCursorContinues() async throws {
        let t = try TestDataDir()
        defer { t.cleanup() }
        let two = #"{"id":"2","created_at":"\#(isoDate(2))","full_text":"two","user":{"username":"alice"},"entities":[{"type":"photo","link":"https://pbs.twimg.com/media/A1.jpg"},{"type":"photo","link":"https://pbs.twimg.com/media/A2.jpg"}]}"#
        let one = #"{"id":"1","created_at":"\#(isoDate(1))","full_text":"one","user":{"username":"alice"},"entities":[{"type":"photo","link":"https://pbs.twimg.com/media/B1.jpg"}]}"#
        try await t.importUser("alice", [two, one])
        let repo = TweetRepository(t.db)

        let page = try await repo.getMediaPage(
            usernames: ["alice"], after: nil, limit: 10, ascending: true)
        // ツイートは古い順、同一ツイート内の idx は昇順のまま
        #expect(page.map { "\($0.tweetId):\($0.idx)" } == ["1:1", "2:1", "2:2"])

        // 3 要素カーソルの継続
        let p1 = try await repo.getMediaPage(
            usernames: ["alice"], after: nil, limit: 2, ascending: true)
        let last = p1.last!
        let p2 = try await repo.getMediaPage(
            usernames: ["alice"],
            after: MediaCursor(sortKey: last.sortKey, idInt: last.idInt, idx: last.idx),
            limit: 2, ascending: true)
        #expect(p1.map { "\($0.tweetId):\($0.idx)" } == ["1:1", "2:1"])
        #expect(p2.map { "\($0.tweetId):\($0.idx)" } == ["2:2"])
    }

    @Test func ascendingCombinesWithRangeAndUnreadOnly() async throws {
        let t = try TestDataDir()
        defer { t.cleanup() }
        try await t.importUser("alice", [
            tweetLine(1, date: isoDate(1)), tweetLine(2, date: isoDate(2)),
            tweetLine(3, date: isoDate(3)), tweetLine(4, date: isoDate(4)),
        ])
        let repo = TweetRepository(t.db)
        try await repo.setRead(tweetId: "3", username: "alice", read: true)

        // 期間 [Day2, Day5) かつ未読のみ → 2, 4 が古い順
        let page = try await repo.getPage(
            usernames: ["alice"], unreadOnly: true, after: nil, limit: 10,
            range: DateRangeFilter(fromEpoch: isoEpoch(2), toEpochExclusive: isoEpoch(5)),
            ascending: true)

        #expect(page.rows.map(\.tweetId) == ["2", "4"])
    }
}

/// 期間フィルタの境界規則 (docs/viewer-features.md §7.3 — ローカル 0 時の半開区間)。
/// タイムゾーン依存は JST 固定で検証する。
@Suite struct DateRangeFilterTests {
    private let jst = TimeZone(identifier: "Asia/Tokyo")!

    @Test func 月境界はローカル0時の半開区間() {
        // JST 2026-07 の月フィルタ: [2026-06-30T15:00Z, 2026-07-31T15:00Z)
        let range = DateRangeFilter.fromLocalParts(year: 2026, month: 7, day: nil, timeZone: jst)
        let junE30_15utc = Int64(1_782_831_600)   // 2026-06-30T15:00:00Z = JST 7/1 0:00
        #expect(range.fromEpoch == junE30_15utc)
        #expect(range.toEpochExclusive == junE30_15utc + 31 * 86_400)
    }

    @Test func 日フィルタは1日幅() {
        let range = DateRangeFilter.fromLocalParts(year: 2026, month: 7, day: 15, timeZone: jst)
        #expect(range.toEpochExclusive - range.fromEpoch == 86_400)
    }

    @Test func 年フィルタは1年幅でうるう年も正しい() {
        let range = DateRangeFilter.fromLocalParts(year: 2024, month: nil, day: nil, timeZone: jst)
        #expect(range.toEpochExclusive - range.fromEpoch == 366 * 86_400)   // 2024 はうるう年
    }

    @Test func yearsCoveredはローカル暦の年() {
        // 2018-12-31T20:00Z は JST では 2019-01-01T05:00
        var comps = DateComponents()
        (comps.year, comps.month, comps.day, comps.hour) = (2018, 12, 31, 20)
        comps.timeZone = TimeZone(identifier: "UTC")
        let epoch = Int64(Calendar(identifier: .gregorian).date(from: comps)!.timeIntervalSince1970)
        #expect(DateRangeFilter.yearsCovered(minEpoch: epoch, maxEpoch: epoch, timeZone: jst) == [2019])
        #expect(DateRangeFilter.yearsCovered(
            minEpoch: epoch - 86_400 * 400, maxEpoch: epoch, timeZone: jst) == [2017, 2018, 2019])
    }

    @Test func daysInは月の実日数() {
        #expect(DateRangeFilter.daysIn(year: 2026, month: 2, timeZone: jst) == 28)
        #expect(DateRangeFilter.daysIn(year: 2024, month: 2, timeZone: jst) == 29)
        #expect(DateRangeFilter.daysIn(year: 2026, month: 7, timeZone: jst) == 31)
    }

    @Test func 月初のJST早朝ツイートが前月に漏れない() async throws {
        // 2026-07-01T05:00 JST = 2026-06-30T20:00Z。7 月フィルタに含まれ、6 月には含まれない
        let t = try TestDataDir()
        defer { t.cleanup() }
        try await t.importUser("alice", [
            tweetLine(1, date: "2026-06-30T20:00:00+00:00"),
        ])
        let repo = TweetRepository(t.db)
        let july = DateRangeFilter.fromLocalParts(year: 2026, month: 7, day: nil, timeZone: jst)
        let june = DateRangeFilter.fromLocalParts(year: 2026, month: 6, day: nil, timeZone: jst)
        let inJuly = try await repo.getPage(
            usernames: ["alice"], unreadOnly: false, after: nil, limit: 10, range: july)
        let inJune = try await repo.getPage(
            usernames: ["alice"], unreadOnly: false, after: nil, limit: 10, range: june)
        #expect(inJuly.rows.count == 1)
        #expect(inJune.rows.isEmpty)
    }
}
